lib/aws-record/record/batch_writer.rb

```diff
@@ -0,0 +1,222 @@
+        validate_items
+        # the -1 is to make it equal the number of retries requested
+        @retry_times = (opts.delete(:retry_count) { |_| 4 } - 1)
+        retry_items = send_chunks(opts)
+        responses = retry_failed_requests(retry_items, opts)
+        analyze_result(responses)
+      end
+
+      # Perform the save! function but capture the error that is raised on
+      # unsuccessful save.
+      def save(opts = {})
+        save!(opts)
+        true
+      rescue RuntimeError
+        false
+      end
+
+      def error?
+        @state == ERROR_SEND_STATE
+      end
+
+      # Get a hash of the responses that didn't work along with their
+      # unprocessed items. Mimics ActiveRecord design.
+      def errors
+        @errors.map do |err|
+          {
+            error: err.error,
+            data: err.data,
+            unprocesssed_items: err.unprocessed_items
+          }
+        end
+      end
+
+      def pending?
+        @state == PENDING_STATE
+      end
+
+      def unprocessed_items
+        # Will only have one error if error is RecordError
+        return [] if @errors.empty? || @errors.first.data.is_a?(Errors::RecordError)
+        # Should be more than one error if failed_items.size > 25
+        @errors.map(&:unprocessed_items)
+      end
+
+      def valid?
+        @state == SUCCESSFUL_SEND_STATE
+      end
+
+      private
+      def alert_errors(responses)
+        @errors += responses.reject(&:successful?)
+        @state = ERROR_SEND_STATE
+        raise Errors::SubmissionError
+      end
+
+      def analyze_result(responses)
+        alert_errors(responses) if failures?(responses)
+        @state = SUCCESSFUL_SEND_STATE
+      end
+
+      def batch_write(item_requests_array, opts)
+        @client.batch_write_item(
+          formatted_request(item_requests_array, opts)
+        )
+      # Retry in instance where client raise_response_error is true
+      rescue Aws::DynamoDB::Errors::ProvisionedThroughputExceededException => e
+        @error_struct.new(e.class, e, item_requests_array)
+      end
+
+      def batch_write_items(item_requests_array, opts)
+        response = batch_write(item_requests_array, opts)
+        if response.successful?
+          response.unprocessed_items[@model.table_name] || []
+        else
+          item_requests_array
+        end
+      end
+
+      def capture_validation_error(e)
+        @state = ERROR_SEND_STATE
+        @errors << @error_struct.new(e.class, e, @items)
+      end
+
+      def failures?(responses)
+        responses.any? do |response|
+          !response.successful? || response.unprocessed_items.empty?
+        end
+      end
+
+      def formatted_request(item_requests_array, opts)
+        {
+          request_items: {
+            @model.table_name => item_requests_array
+          }
+        }.merge(opts)
+      end
+
+      def request_item(item)
+        {
+          put_request: {
+            item: item.build_item_for_save
```

Comment:
There's a bit of a design problem to consider here for batch support. Note that we recently changed our save semantics to prefer `#update_item` when possible. `#batch_write_items` doesn't let us do updates, but we would need to have a story for how to deal with unmodeled attributes getting clobbered, for example.

Some thinking might be required here, which might shelve this for a bit until we capture unmodeled attributes.